Added a cache that holds the recipes for jei compat

Snonky opened the pull request:
### Working on #5 and CyclopsMC/IntegratedDynamics#1247

Made a "tick-wise" cache that saves all incoming JEI recipes and the resulting missing item sets during a single tick to a simple list. When the JEI recipes or the storage contents change the cache clears and repopulates with the recipes of the new tick.

JEI always calls `transferRecipe` for each recipe in the current page in a tick so the cache should never hold more than 4 recipes. But if there is a JEI screen that shows more than 4 recipes it should still work.

____

Videos showcasing the performance gain. The terminal is still pretty hard on the client with a large storage but it's a step ^^

#### Before 

https://user-images.githubusercontent.com/47639380/235269260-de9e7269-6b72-4350-91ce-1b0ef5d9e7ac.mp4

#### After

https://user-images.githubusercontent.com/47639380/235269280-8a6a49fd-df81-4e4f-9c61-88055f3cf296.mp4

Snonky committed:
Added a cache that holds the recipes for jei compat

rubensworks reviewed (changes requested):
Thanks for the PR!

While the caching idea is good, I'm not sure implementing a custom cache is the best way forward.
Can we instead use an existing caching implementation (such as google's cache, [example](https://github.com/CyclopsMC/CyclopsCore/blob/901c71067e565ff38b2fa394c01692d826fb6756/src/main/java/org/cyclops/cyclopscore/datastructure/DimPos.java#L33-L40))?
We probably can cache for a minute or so, and I don't see a need to limit the number of cached recipes.
Cache would still have to be forcefully invalidated when changeId is modified of course.

Could you also target the 1.19-lts (or 1.18) branch here?

Snonky commented:
I deleted my forked branch because I rebased to `master-1.19-lts`. I was unaware that permanently closes the github PR. I opened a new PR for the changes. Sorry for the mess.
